prov/util: add a function to move iov and desc based on given len

a-ilango opened the pull request:
This is intended to be used by rxm and mrail providers.

shefty commented on src/iov.c:
```diff
@@ -101,3 +101,41 @@ int ofi_truncate_iov(struct iovec *iov, size_t *iov_count, size_t trim_size)
+		src->iov[0].iov_len -= len;
+	}
+	return 0;
+}
```
This looks like a merger of memcpy iovecs, ofi_consume_iov(), followed by ofi_truncate_iov().

a-ilango replied:
Yes. In addition, this avoids looping through src->iov again (since consume and truncate are merged into one) and also handles desc.

a-ilango commented:
In rxm, the use case is the rendezvous protocol. The initiating side conveys tx iovec in the protocol header and the receiving side issues an RMA read for each buffer in the remote iovec. The recv iovec that's been posted by the app has to be split based on the remote buffer's length.

> This argues for a slightly different approach

What do you have in mind?

a-ilango commented:
Another approach I was thinking about is to have an "iovec iterator" like function that would give us the next available list of buffers for requested length. Once the source iovec is exhausted it would return an error.

shefty commented:
For multi-rail, I was thinking of something like ofi_divide_iov() -- that would split a single iovec into N iovecs based on length.  At least that seems like the desired semantic.

The rxm case is trickier.  It sounds like it need to take as input 2 iovecs A and B) and create a 3rd iovec that's some sort of merger based on applying the length of iovec B on iovec A.

shefty commented on src/iov.c:
```diff
@@ -101,3 +101,42 @@ int ofi_truncate_iov(struct iovec *iov, size_t *iov_count, size_t trim_size)
+
+		src->iov[0].iov_base = (uint8_t *)src->iov[0].iov_base + len;
+		src->iov[0].iov_len -= len;
+	}
```
Based on our discussion yesterday, I think this fuctionality is close.  Does it matter if you remove 'len' bytes from the front or back of the iovec?  If it doesn't, trimming it off the back might be easier and avoid the memmove adjustments  needed on the src iov.

a-ilango commented:
I've updated the function to not modify the source iovec. It now accepts index and offset parameters to keep track of starting position in source iovec. Also added a split iovec function.

yburette commented:
As discussed with @a-ilango, I was thinking that we could have a copy function that would also hide the bookkeeping of index and offset.

If we make index and offset in/out parameters, then the caller only has to provide storage for them. It is the copy function that updates index and offset.

`ofi_copy_iov_desc(struct ofi_iov_desc *src, size_t len, struct ofi_iov_desc *dst, size_t *index, size_t *offset)`

The caller can then re-use index and offset with any subsequent call to this function.

What do you think?

shefty commented on src/iov.c:
```diff
@@ -101,3 +101,51 @@ int ofi_truncate_iov(struct iovec *iov, size_t *iov_count, size_t trim_size)
+
+/* Copy 'len' bytes worth of src iovec to dst */
+int ofi_copy_iov_desc(struct ofi_iov_desc *src, size_t index, size_t offset,
+		      size_t len, struct ofi_iov_desc *dst)
```
nit: The other iov copy routines have the destination as the first parameter, similar to memcpy.

a-ilango commented:
Updated function signature to have dst as first arg. Changed index and offset to be in/out parameters.

a-ilango commented:
I'm gonna remove the ofi_iov_desc struct since it's not directly usable by callers.
